File: hello-tts-dart/bin/hello_multilingual.dart
#!/usr/bin/env dart

import 'dart:io';
import 'dart:convert';
import 'dart:typed_data';
import 'package:path/path.dart' as path;
import 'package:logging/logging.dart';
import '../lib/hello_tts.dart';

final Logger logger = Logger('multilingual_demo');

/// Language configuration structure
class LanguageConfig {
  final String code;
  final String name;
  final String flag;
  final String text;
  final String voice;
  final String? altVoice;

  LanguageConfig({
    required this.code,
    required this.name,
    required this.flag,
    required this.text,
    required this.voice,
    this.altVoice,
  });

  factory LanguageConfig.fromJson(Map<String, dynamic> json) {
    return LanguageConfig(
      code: json['code'] as String,
      name: json['name'] as String,
      flag: json['flag'] as String,
      text: json['text'] as String,
      voice: json['voice'] as String,
      altVoice: json['alt_voice'] as String?,
    );
  }
}

String backend = Platform.environment['TTS_BACKEND'] ?? 'edge';

/// Load language configuration from JSON file
Future<List<LanguageConfig>> loadLanguageConfig(String backend) async {
  // Prefer unified tts_config.json if present (keeps parity with the shell script)
  final unifiedPath = '../shared/tts_config.json';

  try {
    File file;
    if (await File(unifiedPath).exists()) {
      file = File(unifiedPath);
      final content = await file.readAsString();
      final Map<String, dynamic> config = json.decode(content);
      final List<dynamic> languagesJson = config['languages'] ?? [];
      // tts_config.json uses 'google_voice'/'edge_voice' fields; normalize to 'voice'
      return languagesJson
          .map((v) {
            final map = v as Map<String, dynamic>;
            final voice = (backend.toLowerCase() == 'google')
                ? (map['google_voice'] ?? map['voice'] ?? '')
                : (map['edge_voice'] ?? map['voice'] ?? '');
            return LanguageConfig(
              code: map['code'] ?? '',
              name: map['name'] ?? '',
              flag: map['flag'] ?? '',
              text: map['text'] ?? '',
              voice: voice,
              altVoice: map['alt_voice'] ?? '',
            );
          })
          .where((lc) => lc.code.isNotEmpty)
          .toList();
    } else {
      logger.severe('❌ Configuration file not found (tried $unifiedPath');
      return [];
    }
  } catch (e) {
    logger.severe('❌ Error loading configuration: $e');
    return [];
  }
}

/// Create output directory if it doesn't exist
Future<void> createOutputDirectory(String directory) async {
  final dir = Directory(directory);
  if (!await dir.exists()) {
    await dir.create(recursive: true);
  }
}

/// Generate audio for a single language
Future<bool> generateAudioForLanguage(
  TTSProcessor client,
  LanguageConfig languageConfig,
  String outputDir, {
  bool playAudio = false,
}) async {
  final langCode = languageConfig.code;
  final langName = languageConfig.name;
  final flag = languageConfig.flag;
  final text = languageConfig.text;
  final voice = languageConfig.voice;
  final altVoice = languageConfig.altVoice;

  logger.info('\n$flag $langName (${langCode.toUpperCase()})');
  logger.info('Text: $text');
  logger.info('Voice: $voice');

  try {
    // Try primary voice first
    Uint8List? audioData;
    String usedVoice = voice;

    try {
      audioData = await client.synthesizeText(text, voice);
    } catch (e) {
      logger.warning('Primary voice failed: $e');
      if (altVoice != null) {
        logger.info('Trying alternative voice: $altVoice');
        try {
          audioData = await client.synthesizeText(text, altVoice);
          usedVoice = altVoice;
        } catch (e2) {
          logger.warning('Alternative voice also failed: $e2');
          rethrow;
        }
      } else {
        rethrow;
      }
    }

    // Generate filename
    final timestamp = DateTime.now().millisecondsSinceEpoch ~/ 1000;
    final langPrefix = langCode.split('-')[0]; // e.g., 'zh' from 'zh-cn'
    final filename = '${langPrefix}_dart_${backend}_$timestamp.mp3';
    final outputPath = path.join(outputDir, filename);

    // Save audio
    final file = File(outputPath);
    await file.writeAsBytes(audioData);

    logger.info('✅ Generated: $filename');
    logger.info('📁 Saved to: $outputPath');
    logger.info('🎤 Used voice: $usedVoice');

    // Play audio if requested
    if (playAudio) {
      try {
        logger.info('🔊 Playing audio...');
        final player = AudioPlayer();
        await player.play(outputPath);
        logger.info('✅ Playback completed');
      } catch (e) {
        logger.warning('⚠️  Could not play audio: $e');
      }
    }

    return true;
  } catch (e) {
    logger.severe('❌ Failed to generate audio for $langName: $e');
    return false;
  }
}

/// Main function for multilingual demo
Future<void> main(List<String> args) async {
  // Configure logging
  Logger.root.level = Level.INFO;
  Logger.root.onRecord.listen((record) {
    print('${record.level.name}: ${record.message}');
  });

  if (args.isNotEmpty) {
    for (int i = 0; i < args.length; i++) {
      if (args[i] == '--backend' && i + 1 < args.length) {
        backend = args[i + 1];
        break;
      }
    }
  }

  final languages = await loadLanguageConfig(backend);
  if (languages.isEmpty) {
    logger.severe('❌ Failed to load language configuration or no engine found');
    exit(1);
  }

  logger.info('📋 Found ${languages.length} languages to process');

  // Create output directory
  const outputDir = 'output';
  await createOutputDirectory(outputDir);
  final outputPath = path.absolute(outputDir);
  logger.info('📁 Output directory: $outputPath');

  // Initialize TTS client
  late TTSProcessor client;
  try {
    client = TTSProcessor(backend: backend);
    logger.info('✅ TTS client initialized with $backend backend');
  } catch (e) {
    logger.severe('❌ Failed to initialize TTS client: $e');
    exit(1);
  }

  // Process each language
  int successfulCount = 0;
  int failedCount = 0;
  final startTime = DateTime.now();

  for (int i = 0; i < languages.length; i++) {
    final languageConfig = languages[i];
    logger.info('\n📍 Processing language ${i + 1}/${languages.length}');

    final success = await generateAudioForLanguage(
      client,
      languageConfig,
      outputDir,
      playAudio: false, // Set to true if you want to play each audio
    );

    if (success) {
      successfulCount++;
    } else {
      failedCount++;
    }

    // Small delay between languages to be polite to the service
    if (i < languages.length - 1) {
      logger.info('⏳ Waiting before next language...');
      await Future.delayed(const Duration(seconds: 2));
    }
  }

  // Summary
  final endTime = DateTime.now();
  final duration = endTime.difference(startTime);

  logger.info('\n🏁 Processing Complete!');
  logger.info('=' * 40);
  logger.info('✅ Successful: $successfulCount');
  logger.info('❌ Failed: $failedCount');
  logger.info(
      '⏱️  Total time: ${duration.inSeconds}.${duration.inMilliseconds % 1000} seconds');
  logger.info('📁 Output files saved in: $outputPath');

  if (successfulCount > 0) {
    logger.info(
        '\n🎉 Successfully generated audio files for $successfulCount languages!');
    logger
        .info('You can find all generated MP3 files in the output directory.');
  }

  exit(failedCount == 0 ? 0 : 1);
}

File: hello-tts-python/hello_multilingual.py
#!/usr/bin/env python3
"""
Multilingual batch demo for TTS Python implementation.
Generates audio files for multiple languages with specified sentences.
"""

from audio_player import AudioPlayer, AudioError
from utils import create_output_directory
import asyncio
import json
import os
import time
import logging
import sys
from dataclasses import dataclass
from typing import Optional, Dict, List, Any

# Add the parent directory to Python path to import local modules
sys.path.append(os.path.dirname(os.path.abspath(__file__)))

# Configure logging
logger = logging.getLogger('multilingual_demo')
handler = logging.StreamHandler()
handler.setFormatter(logging.Formatter('%(levelname)s: %(message)s'))
logger.addHandler(handler)
logger.setLevel(logging.INFO)

# Dynamically import TTSClient and TTSError based on backend


def get_tts_client_and_error(backend):
    if backend == 'google':
        from backends.google.tts_client import TTSClient, TTSError
    else:
        from backends.edge.tts_client import TTSClient, TTSError
    return TTSClient, TTSError


@dataclass
class LanguageConfig:
    """Language configuration structure"""
    code: str
    name: str
    flag: str
    text: str
    voice: str
    alt_voice: Optional[str] = None

    @classmethod
    def from_dict(cls, data: Dict[str, Any], backend: str) -> 'LanguageConfig':
        voice = data[f'{backend}_voice'] if f'{backend}_voice' in data else data.get(
            'voice', '')
        return cls(
            code=data.get('code', ''),
            name=data.get('name', ''),
            flag=data.get('flag', ''),
            text=data.get('text', ''),
            voice=voice,
            alt_voice=data.get('alt_voice')
        )


async def load_language_config(backend: str) -> List[LanguageConfig]:
    """Load language configuration from shared config file."""
    config_path = "../shared/tts_config.json"
    try:
        with open(config_path, 'r', encoding='utf-8') as f:
            config_json = json.load(f)
            languages_data = config_json.get('languages', [])

            return [
                LanguageConfig.from_dict(lang_data, backend)
                for lang_data in languages_data
                if lang_data.get('code')  # Filter out entries without a code
            ]

    except FileNotFoundError:
        logger.error(f"❌ Configuration file not found: {config_path}")
        return []
    except json.JSONDecodeError as e:
        logger.error(f"❌ Error parsing configuration file: {e}")
        return []
    except Exception as e:
        logger.error(f"❌ Error loading configuration: {e}")
        return []


async def generate_audio_for_language(client, language_config: LanguageConfig, output_dir: str, backend: str, play_audio: bool = False) -> bool:
    """Generate audio for a single language configuration."""
    lang_code = language_config.code
    lang_name = language_config.name
    flag = language_config.flag
    text = language_config.text
    voice = language_config.voice
    alt_voice = language_config.alt_voice

    logger.info(f"\n{flag} {lang_name} ({lang_code.upper()})")
    logger.info(f"Text: {text}")
    logger.info(f"Voice: {voice}")

    try:
        # Try primary voice first
        audio_data = None
        used_voice = voice

        import inspect
        try:
            synthesize = client.synthesize_text
            result = synthesize(text, voice)

            # Handle both async and sync results
            if inspect.isawaitable(result):
                audio_data = await result
            else:
                audio_data = result

            if audio_data is None:
                logger.error(
                    f"❌ Synthesis failed for {lang_name}: No audio returned")
                return False

        except Exception as e:
            logger.warning(f"Primary voice failed: {e}")
            if alt_voice:
                logger.info(f"Trying alternative voice: {alt_voice}")
                try:
                    result = synthesize(text, alt_voice)

                    # Handle both async and sync results for alt voice
                    if inspect.isawaitable(result):
                        audio_data = await result
                    else:
                        audio_data = result

                    if audio_data is None:
                        logger.error(
                            f"❌ Synthesis failed for {lang_name} (alt voice): No audio returned")
                        return False

                    used_voice = alt_voice
                except Exception as e2:
                    logger.warning(f"Alternative voice also failed: {e2}")
                    raise e2
            else:
                raise e

        # Generate filename
        timestamp = int(time.time())
        lang_prefix = lang_code.split('-')[0]  # e.g., 'zh' from 'zh-cn'
        backend_name = 'gtts' if backend == 'google' else backend
        filename = f"{lang_prefix}_python_{backend_name}_{timestamp}.mp3"
        output_path = os.path.join(output_dir, filename)

        # Save audio (handle both async and sync save_audio methods)
        save_audio = client.save_audio
        if inspect.iscoroutinefunction(save_audio):
            await save_audio(audio_data, output_path)
        else:
            save_audio(audio_data, output_path)

        logger.info(f"✅ Generated: {filename}")
        logger.info(f"📁 Saved to: {os.path.abspath(output_path)}")
        logger.info(f"🎤 Used voice: {used_voice}")

        # Play audio if requested
        if play_audio:
            try:
                logger.info("🔊 Playing audio...")
                player = AudioPlayer()
                player.play_file(output_path)
                logger.info("✅ Playback completed")
            except AudioError as e:
                logger.warning(f"⚠️  Could not play audio: {e}")

        return True

    except Exception as e:
        logger.error(f"❌ Failed to generate audio for {lang_name}: {e}")
        return False


async def main() -> int:
    """Main function for multilingual demo."""
    logger.info("🌍 Multilingual TTS Demo - Python Implementation")
    logger.info("=" * 60)

    # Determine backend from environment or arguments
    backend = os.environ.get('TTS_BACKEND')
    if not backend:
        for i, arg in enumerate(sys.argv):
            if arg == '--backend' and i + 1 < len(sys.argv):
                backend = sys.argv[i + 1]
                break
    backend = backend or 'edge'

    # Load language configuration
    languages = await load_language_config(backend)
    if not languages:
        logger.error(
            "❌ Failed to load language configuration or no languages found")
        return 1

    logger.info(f"📋 Found {len(languages)} languages to process")

    # Create output directory
    output_dir = "output"
    create_output_directory(output_dir)
    logger.info(f"📁 Output directory: {os.path.abspath(output_dir)}")

    # Initialize TTS client
    try:
        TTSClient, TTSError = get_tts_client_and_error(backend)
        client = TTSClient()
        logger.info(f"✅ TTS client initialized with {backend} backend")
    except Exception as e:
        logger.error(f"❌ Failed to initialize TTS client: {e}")
        return 1

    # Process each language
    successful_count = 0
    failed_count = 0
    start_time = time.time()

    for i, language_config in enumerate(languages):
        logger.info(f"\n📍 Processing language {i + 1}/{len(languages)}")

        success = await generate_audio_for_language(
            client,
            language_config,
            output_dir,
            backend,
            play_audio=False  # Set to True if you want to play each audio
        )

        if success:
            successful_count += 1
        else:
            failed_count += 1

        # Small delay between languages to be polite to the service
        if i < len(languages) - 1:
            logger.info("⏳ Waiting before next language...")
            await asyncio.sleep(2)

    # Summary
    end_time = time.time()
    duration = end_time - start_time

    logger.info("\n🏁 Processing Complete!")
    logger.info("=" * 40)
    logger.info(f"✅ Successful: {successful_count}")
    logger.info(f"❌ Failed: {failed_count}")
    logger.info(f"⏱️  Total time: {duration:.2f} seconds")
    logger.info(f"📁 Output files saved in: {os.path.abspath(output_dir)}")

    if successful_count > 0:
        logger.info(
            f"\n🎉 Successfully generated audio files for {successful_count} languages!")
        logger.info(
            "You can find all generated MP3 files in the output directory.")

    return 0 if failed_count == 0 else 1


if __name__ == "__main__":
    try:
        exit_code = asyncio.run(main())
        sys.exit(exit_code)
    except KeyboardInterrupt:
        print("\n⏹️  Operation cancelled by user")
        sys.exit(1)
    except Exception as e:
        print(f"\n💥 Unexpected error: {e}")
        sys.exit(1)
